### PR TITLE
MODINREACH-21 D2IR Local Endpoint (Circulation): Patron Verification Endpoint in mod-inn-reach

### PR DESCRIPTION
**Pull request by oliinyko**
## Purpose
The INN-Reach central server should be able to verify that the requesting user is a valid patron of the library they claim to be.

https://issues.folio.org/browse/MODINREACH-21

## Approach
- Add patron verify controller
- Add possibility to load user by publicId (barcode or externalSystemId)
- Add patron and patron-blocks clients to load required info
- Add tests

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/innreach/domain/service/impl/PatronInfoServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  private final PatronClient patronClient;
+  private final PatronBlocksClient patronBlocksClient;
+  private final InnReachResponseMapper mapper;
+
+  @Override
+  public PatronInfoResponseDTO verifyPatron(String centralServerCode, String visiblePatronId,
+                                            String patronAgencyCode, String patronName) {
+    PatronInfoResponse response;
+    try {
+      var centralServerId = centralServerService.getCentralServerByCentralCode(centralServerCode).getId();
```
Comment (andrew-khodyka) [inline src/main/java/org/folio/innreach/domain/service/impl/PatronInfoServiceImpl.java]: I think this isn't the best method to get centralServer ID, because loads too much data with JOIN FETCH statements. Maybe we could add a separate method, something like "getCentralServerIdByCentralCode", what do you think?

File: src/main/java/org/folio/innreach/domain/service/impl/PatronInfoServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    var blocks = patronBlocksClient.getPatronBlocks(user.getId()).getResult();
+
+    if (CollectionUtils.isNotEmpty(blocks)) {
+      var block = blocks.stream()
+        .filter(PatronBlock::getBlockBorrowing)
+        .filter(PatronBlock::getBlockRequests)
+        .findFirst()
+        .orElse(null);
+
+      Assert.isTrue(block == null, "Patron block found: " + block.getMessage());
```
Comment (andrew-khodyka) [inline src/main/java/org/folio/innreach/domain/service/impl/PatronInfoServiceImpl.java]: Is this assert correct? block == null -> "Patron block found" ?

Comment (oliinyko) [reply src/main/java/org/folio/innreach/domain/service/impl/PatronInfoServiceImpl.java]: Yes, in this case we expect no blocks and if any is found then it is an exceptional state.
I agree this might be a bit confusing, I will replace this with if condition

File: src/main/java/org/folio/innreach/domain/service/impl/CentralPatronTypeMappingServiceImpl.java [hunk 3]
```diff
@@ -38,6 +38,14 @@ public CentralPatronTypeMappingsDTO getCentralPatronTypeMappings(UUID centralSer
     return mapper.toDTOCollection(centralPatronTypeMappings);
   }
 
+  @Transactional(readOnly = true)
+  @Override
+  public Integer getCentralPatronType(UUID centralServerId, String barcode) {
+    return repository.findOneByCentralServerIdAndBarcode(centralServerId, barcode)
+      .map(CentralPatronTypeMapping::getCentralPatronType)
+      .orElse(null);
```
Comment (dmtkachenko) [inline src/main/java/org/folio/innreach/domain/service/impl/CentralPatronTypeMappingServiceImpl.java]: Optional may be? Or exception, if the type is required according to business logic?

File: src/main/java/org/folio/innreach/domain/service/impl/PatronTypeMappingServiceImpl.java [hunk 4]
```diff
@@ -51,6 +52,13 @@ public PatronTypeMappingsDTO updateAllMappings(UUID centralServerId, PatronTypeM
     return mapper.toDTOCollection(saved);
   }
 
+  @Override
+  public Integer getCentralPatronType(UUID centralServerId, UUID patronGroupId) {
+    return repository.findOneByCentralServerIdAndPatronGroupId(centralServerId, patronGroupId)
+      .map(CentralPatronTypeMapping::getCentralPatronType)
+      .orElse(null);
```
Comment (dmtkachenko) [inline src/main/java/org/folio/innreach/domain/service/impl/PatronTypeMappingServiceImpl.java]: Same here, regarding Optional<Integer> result

File: src/test/java/org/folio/innreach/domain/service/impl/PatronInfoServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+import org.folio.innreach.domain.service.CentralPatronTypeMappingService;
+import org.folio.innreach.domain.service.CentralServerService;
+import org.folio.innreach.domain.service.InnReachTransactionService;
+import org.folio.innreach.domain.service.PatronTypeMappingService;
+import org.folio.innreach.external.mapper.InnReachResponseMapper;
+import org.folio.innreach.external.mapper.InnReachResponseMapperImpl;
+import org.folio.innreach.fixture.CentralServerFixture;
+
+@ExtendWith(MockitoExtension.class)
+class PatronInfoServiceImplTest {
```
Comment (dmtkachenko) [inline src/test/java/org/folio/innreach/domain/service/impl/PatronInfoServiceImplTest.java:44]: please add tests for all negative cases

**Review by dmtkachenko — APPROVED** [review]
